src/lib.rs

```diff
@@ -582,6 +616,121 @@ impl<'a, T: Ord + Clone + 'a> Extend<&'a T> for MinMaxHeap<T> {
     }
 }
 
+/// Structure wrapping a mutable reference to the minumum item on a
+/// `MinMaxHeap`.
+///
+/// This `struct` is created by the [`peek_min_mut`] method on [`MinMaxHeap`]. See
+/// its documentation for more.
+///
+/// [`peek_min_mut`]: struct.MinMaxHeap.html#method.peek_min_mut
+/// [`MinMaxHeap`]: struct.MinMaxHeap.html
+pub struct PeekMinMut<'a, T: 'a + Ord> {
+    heap: &'a mut MinMaxHeap<T>,
+    sift: bool,
+}
+
+impl<T: Ord + fmt::Debug> fmt::Debug for PeekMinMut<'_, T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_tuple("PeekMinMut")
+         .field(&self.heap.0[0])
+         .finish()
+    }
+}
+
+impl<T: Ord> Drop for PeekMinMut<'_, T> {
+    fn drop(&mut self) {
+        if self.sift {
+            self.heap.trickle_down_min(0);
+        }
+    }
+}
+
+impl<T: Ord> Deref for PeekMinMut<'_, T> {
+    type Target = T;
+    fn deref(&self) -> &T {
+        debug_assert!(!self.heap.is_empty());
+        // SAFE: PeekMinMut is only instantiated for non-empty heaps
+        unsafe { self.heap.0.get_unchecked(0) }
+    }
+}
+
+impl<T: Ord> DerefMut for PeekMinMut<'_, T> {
+    fn deref_mut(&mut self) -> &mut T {
+        debug_assert!(!self.heap.is_empty());
+        // SAFE: PeekMinMut is only instantiated for non-empty heaps
+        unsafe { self.heap.0.get_unchecked_mut(0) }
+    }
+}
+
+impl<'a, T: Ord> PeekMinMut<'a, T> {
+    /// Removes the peeked value from the heap and returns it.
+    pub fn pop(mut self) -> T {
```

Comment:
In BinaryHeap, the pop function here takes a parameter called `this` instead of `self`, which means it has to be called with `PeekMut::pop`. It's not clear from the docs why this is... I guess maybe because otherwise `heap.peek_mut().unwrap().pop()` might be ambiguous, like if you have a heap full of vectors or something? Still, it made it pretty clunky to use in practice, so I went with the more typical version here. Happy to change it back though if you feel strongly.

Reply:
This is because `PeekMut` implements `Deref`, so it "inherits" all methods from `T`. If `T` also has a `pop` method then calling `pop` on the `PeekMut` instance becomes ambiguous. This is also why `ManuallyDrop::{drop, into_inner}` both take a parameter named `slot` rather than `self`.